### ISSUE TITLE
Reduction codegen

### ISSUE THREAD

**Issue by wok1909** [PSAL-POSTECH/PyTorchSim#137]
The functionally incorrect MLIR code can be reproduced with second kernel of softmax, input size [16, 256, 256] with the commit https://github.com/PSAL-POSTECH/PyTorchSim/commit/34909a4825a8d4eb3f7ee01195bbb1ac64fbef11.

The correct MLIR code was produced by commenting line 463-464, 469-471 in mlir_common.py.

Functionally correct MLIR code
```
#map0 = affine_map<(index0, index1)[] -> (256*index0 + index1)>
memref.global @arg0_1_0_spad : memref<64x256xf32, 1>
#map1 = affine_map<(d0,d1)[] -> (d0)>
memref.global @buf0_0_spad : memref<64x256xf32, 1>
memref.global @buf1_0_spad : memref<64xf32, 1>
func.func @kernel(%in_ptr0: memref<1048576xf32>,
                       %in_ptr1: memref<4096xf32>,
                       %out_ptr0: memref<4096xf32>)
{
    %const0 = arith.constant 0 : index
    %const1 = arith.constant 8 : index
    %const2 = arith.constant 2 : index
    %const3 = arith.constant 3 : index
    %alloc0 = memref.alloc() : memref<1xi32>
    %spad0 = memref.get_global @arg0_1_0_spad : memref<64x256xf32, 1>
    %spad1 = memref.get_global @buf0_0_spad : memref<64x256xf32, 1>
    %spad2 = memref.get_global @buf1_0_spad : memref<64xf32, 1>
    affine.for %index0 = 0 to 4096 step 64
    {
        %tmp4 = arith.constant 0.0 : f32
        %tmp5 = vector.broadcast %tmp4 : f32 to vector<64xf32>
        %tmp_acc0 = affine.for %index1 = 0 to 256 step 256 iter_args(%iter0 = %tmp5) -> (vector<64xf32>)
        {
            %apply0 = affine.apply #map0(%index0, %index1)[]
            %apply1 = affine.apply #map1(%index0,%const0)[]
            memref.dma_start %in_ptr0[%apply0], %spad0[%const0,%const0], %const2, %alloc0[%const0], %const0, %const1 : memref<1048576xf32>, memref<64x256xf32, 1>, memref<1xi32> {padding=1, sram_stride=[1, 64]}
            memref.dma_start %in_ptr1[%apply1], %spad1[%const0,%const0], %const2, %alloc0[%const0], %const0, %const1 : memref<4096xf32>, memref<64x256xf32, 1>, memref<1xi32> {padding=1, sram_stride=[1, 64]}
            %tmp_acc1 = affine.for %compute_idx = 0 to 2048 step 64 iter_args(%iter1 = %iter0) -> (vector<64xf32>)
            {
                %tmp0 = affine.vector_load %spad0[%const0,%compute_idx] : memref<64x256xf32, 1>, vector<64xf32>
                %tmp1 = affine.vector_load %spad1[%const0,%compute_idx] : memref<64x256xf32, 1>, vector<64xf32>
                %tmp2 = arith.subf %tmp0, %tmp1 : vector<64xf32>
                %tmp3 = math.exp %tmp2 : vector<64xf32>
                %tmp6 = arith.addf %tmp3, %iter1 : vector<64xf32>
                affine.yield %tmp6 : vector<64xf32>
            } {inner_loop=false}
            affine.yield %tmp_acc1 : vector<64xf32>
        } {accumulation_loop=true}
        %tmp7 = vector.broadcast %tmp4 : f32 to vector<8xf32>
        %tmp8 = vector.shape_cast %tmp_acc0 : vector<64xf32> to vector<8x8xf32>
        %tmp9 = vector.multi_reduction <add>, %tmp8, %tmp7 [0] : vector<8x8xf32> to vector<8xf32>
        affine.vector_store %tmp9, %spad2[%const0] : memref<64xf32, 1>, vector<8xf32>
        memref.dma_start %spad2[%const0], %out_ptr0[%index0], %const3, %alloc0[%const0], %const0, %const1 : memref<64xf32, 1>, memref<4096xf32>, memref<1xi32> {padding=0, sram_stride=[1]}
    } {outer_loop=true}
    return
}
```

Functionally incorrect MLIR code
```
#map0 = affine_map<(index0, index1)[] -> (256*index0 + index1)>
memref.global @arg0_1_0_spad : memref<2048x8xf32, 1>
#map1 = affine_map<(d0,d1)[] -> (d0)>
memref.global @buf0_0_spad : memref<2048x8xf32, 1>
memref.global @buf1_0_spad : memref<2048xf32, 1>
func.func @kernel(%in_ptr0: memref<1048576xf32>,
                       %in_ptr1: memref<4096xf32>,
                       %out_ptr0: memref<4096xf32>)
{
    %const0 = arith.constant 0 : index
    %const1 = arith.constant 8 : index
    %const2 = arith.constant 2 : index
    %const3 = arith.constant 3 : index
    %alloc0 = memref.alloc() : memref<1xi32>
    %spad0 = memref.get_global @arg0_1_0_spad : memref<2048x8xf32, 1>
    %spad1 = memref.get_global @buf0_0_spad : memref<2048x8xf32, 1>
    %spad2 = memref.get_global @buf1_0_spad : memref<2048xf32, 1>
    affine.for %index0 = 0 to 4096 step 2048
    {
        %tmp4 = arith.constant 0.0 : f32
        %tmp5 = vector.broadcast %tmp4 : f32 to vector<2048xf32>
        %tmp_acc0 = affine.for %index1 = 0 to 256 step 8 iter_args(%iter0 = %tmp5) -> (vector<2048xf32>)
        {
            %apply0 = affine.apply #map0(%index0, %index1)[]
            %apply1 = affine.apply #map1(%index0,%const0)[]
            memref.dma_start %in_ptr0[%apply0], %spad0[%const0,%const0], %const2, %alloc0[%const0], %const0, %const1 : memref<1048576xf32>, memref<2048x8xf32, 1>, memref<1xi32> {padding=1, sram_stride=[1, 2048]}
            memref.dma_start %in_ptr1[%apply1], %spad1[%const0,%const0], %const2, %alloc0[%const0], %const0, %const1 : memref<4096xf32>, memref<2048x8xf32, 1>, memref<1xi32> {padding=1, sram_stride=[1, 2048]}
            %tmp_acc1 = affine.for %compute_idx = 0 to 2048 step 2048 iter_args(%iter1 = %iter0) -> (vector<2048xf32>)
            {
                %tmp0 = affine.vector_load %spad0[%const0,%compute_idx] : memref<2048x8xf32, 1>, vector<2048xf32>
                %tmp1 = affine.vector_load %spad1[%const0,%compute_idx] : memref<2048x8xf32, 1>, vector<2048xf32>
                %tmp2 = arith.subf %tmp0, %tmp1 : vector<2048xf32>
                %tmp3 = math.exp %tmp2 : vector<2048xf32>
                %tmp6 = arith.addf %tmp3, %iter1 : vector<2048xf32>
                affine.yield %tmp6 : vector<2048xf32>
            } {inner_loop=false}
            affine.yield %tmp_acc1 : vector<2048xf32>
        } {accumulation_loop=true}
        %tmp7 = vector.broadcast %tmp4 : f32 to vector<256xf32>
        %tmp8 = vector.shape_cast %tmp_acc0 : vector<2048xf32> to vector<8x256xf32>
        %tmp9 = vector.multi_reduction <add>, %tmp8, %tmp7 [0] : vector<8x256xf32> to vector<256xf32>
        affine.vector_store %tmp9, %spad2[%const0] : memref<2048xf32, 1>, vector<256xf32>
        memref.dma_start %spad2[%const0], %out_ptr0[%index0], %const3, %alloc0[%const0], %const0, %const1 : memref<2048xf32, 1>, memref<4096xf32>, memref<1xi32> {padding=0, sram_stride=[1]}
    } {outer_loop=true}
    return
}
```
